### ISSUE TITLE
workspace schedule: If I update my workspace, the autostop resets to updated_at + 12 hours

### ISSUE THREAD

**Issue by matifali** [coder/coder#8546]
I have my workspace to start daily at 8 AM and then autostop after 12 hours (8 PM).
I updated my workspace at 2:49 PM with a new template version, and now my autostop time shows 2:49 AM.

### Expected
Workspace should follow the original autostop schedule 12 hours after autostart (18 PM).

### Actual

Workspace says
![image](https://github.com/coder/coder/assets/10648092/fd69912a-a61b-4c1f-a89e-ef9634b1da7a)

**Comment by johnstcn** [comment]
@matifali can you clarify the impact of this issue? Is it just confusing, or is there a bigger knock-on effect that I'm missing here?

**Comment by johnstcn** [comment]
@matifali Should restarting the workspace in the middle of the day also not push the auto-stop deadline further? What about stopping the workspace, waiting 1 minute, and starting the workspace again?

My main question is:- should a start within the autostop window of a previous start honour the previous autostop window?

**Comment by matifali** [comment]
> @matifali Should restarting the workspace in the middle of the day also not push the auto-stop deadline further? What about stopping the workspace, waiting 1 minute, and starting the workspace again?

Yes

> My main question is:- should a start within the autostop window of a previous start honour the previous autostop window?

Yes

Also @bpmct can add further comments. Maybe I misunderstood the expected behavior. My understanding is that autostart/autostop are for cost control and should be followed strictly unless the workspace is active at autostop time.(In that case we automatically bump the shutdown time) See: #2995

**Comment by johnstcn** [comment]
@matifali OK, so just to make sure we're both on the same page:

Say a user starts a workspace at $T_{start}$, and stops it manually at $T_{stop}$, there would exist a window of time between $T_{stop}$ and $T_{autostop}$.

If a user then starts their workspace again at $U_{start}$ such that $T_{stop} < U_{start} < T_{autostop}$, then $U_{autostop}$ should equal $T_{autostop}$. Is that correct?

```
                         | <-- starts in this window should  --> |
                         |     re-use the old autostop time
   |=====================|................|......................|
   T_{start}             T_{stop}         U_{start}              T_{autostop}
```

**Comment by matifali** [comment]
> @matifali OK, so just to make sure we're both on the same page:
> 
> Say a user starts a workspace at Tstart, and stops it manually at Tstop, there would exist a window of time between Tstop and Tautostop.
> 
> If a user then starts their workspace again at Ustart such that Tstop<Ustart<Tautostop, then Uautostop should equal Tautostop. Is that correct?
> 
> ```
>                          | <-- starts in this window should  --> |
>                          |     re-use the old autostop time      |
>    |=====================|................|......................|
>    T_{start}             T_{stop}         U_{start}              T_{autostop}
> ```

@johnstcn Yes, exactly. This is my understanding of the expected behavior.

**Comment by matifali** [comment]
I gave it a 2nd thought and there can be 2 situations. 

1. In an enterprise environment where the workday starts exactly at 8 AM and the developer isn't allowed to keep the workspace running for more than 12 hours. They can set an auto-start at 8 AM and then automatically auto-stop at 8 PM after 12 hours. It should ignore any updates and restarts during the workday. 
2. In a remote working setup where developers follow a flexible schedule, they should just disable auto-start but can enable auto-stop to be 12 hours + the latest start/restart/update time. They can also totally turn off auto stop as developers can stop when they want to consider a flexible setup. 

In short. For autostop, we can keep our current behavior when there is no auto-start set but should strictly follow autostart time when it is set.

**Comment by johnstcn** [comment]
1) sounds like a use-case for quotas (https://coder.com/docs/v2/latest/admin/quotas)

**Comment by deansheather** [comment]
The autostop TTL is the amount of time of inactivity before the workspace is autostopped. It's not related to the autostart time at all.

#8115 reworks max_ttl on templates and doesn't impact the inactivity TTL.

**Comment by matifali** [comment]
@deansheather, I could not understand how this is going to be addressed.
